File: 139.word-break.cpp
/*
 * @lc app=leetcode id=139 lang=cpp
 *
 * [139] Word Break
 */
#include <bits/stdc++.h>

using namespace std;

// @lc code=start
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        set<string> wordDictSet;
        for (string word: wordDict) {
            wordDictSet.insert(word);
        }

        int n = s.size();

        vector<vector<int>> dp(n+1, vector<int>(n+1, -1));
        
        return TopDown(s, wordDictSet, 0, n-1, dp)==1 ? true : false;
    }

    int TopDown(string s, set<string>& wordDictSet, int start, int end, vector<vector<int>>& dp) {
        if (dp[start][end] != -1) { return dp[start][end]; }
        
        // Check if s[start...end] is present in wordDictSet
        int length = end - start + 1;
        string substring = s.substr(start, length);
        if (wordDictSet.find(substring) != wordDictSet.end()) {
            dp[start][end] = true;
            wordDictSet.insert(substring);
            return dp[start][end];
        }

        dp[start][end] = false; 
        for (int middle=start; middle<end; middle++) {
            bool answer = TopDown(s, wordDictSet, start, middle, dp) && TopDown(s, wordDictSet, middle+1, end, dp);
            if (answer) {
                dp[answer][end] = true;
                return true;
            }
        }
        return dp[start][end];
    }

};
// @lc code=end

File: 139.word-break.py
#
# @lc app=leetcode id=139 lang=python3
#
# [139] Word Break
#
from typing import List

# @lc code=start
class TrieNode:
    def __init__(self):
        self.nodes = [None] * 26
        self.EOW = False

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word: str):
        curr = self.root
        for char in word:
            idx = ord(char) - ord('a')
            if curr.nodes[idx] is None:
                curr.nodes[idx] = TrieNode()
            curr = curr.nodes[idx]
        curr.EOW = True
    
    def search(self, word: str) -> int:
        curr = self.root
        dist = 0
        for char in word:
            idx = ord(char) - ord('a')
            if curr.nodes[idx] is None:
                break
            curr = curr.nodes[idx]
            dist += 1
        return dist if curr.EOW else 0

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        trie = Trie()
        for word in wordDict:
            trie.insert(word)
        i = 0
        while i < len(s):
            dist = trie.search(s[i:])
            if dist == 0:
                return False
            i += dist
        return True
             
# @lc code=end

